### PR TITLE
ELEMENTS-2091: Clear singleton mechanical SonarCloud findings [maintenance-3.1.x]

### PR DESCRIPTION
**Pull request by naveen-konda**
> **Companion PR:** same change on `lts-2025` — #1688

## Problem

SonarCloud reports 11 singleton maintainability findings across 7 rules, spread over 9 files. Individually each is too small for its own ticket, so they are grouped here. Every one is a local rewrite of a single expression.

Jira: [ELEMENTS-2091](https://hyland.atlassian.net/browse/ELEMENTS-2091)

## Changes

| Rule | Site | Change |
| --- | --- | --- |
| `S4624` ×2 | `nuxeo-i18n-behavior.js`, `nuxeo-routing-behavior.js` | Nested template literals extracted into a named intermediate (`languageSuffix`, `base`/`separator`). |
| `S4043` ×2 | `nuxeo-aggregation-behavior.js`, `nuxeo-document-acl-table.js` | Copy before mutating: `labels.slice().reverse()`, `[...aces].sort(…)`. |
| `S7741` ×2 | `iron-data-table.js`, `nuxeo-filters-behavior.js` | `typeof x === 'undefined'` → `x === undefined`. |
| `S7766` ×2 | `nuxeo-page-provider-display-behavior.js` | Shift-select range ternaries → `Math.min()` / `Math.max()`. |
| `S7740` ×1 | `nuxeo-user-tag.js` | Dropped the `const selectedElement = this` alias. |
| `S7786` ×1 | `nuxeo-routing-behavior.js` | The type-validation `throw` uses `TypeError` instead of `Error`. |
| `S7753` ×1 | `nuxeo-actions-menu.js` | `path.findIndex((el) => el === action)` → `path.indexOf(action)`. |

Ten source files appear in the diff. `nuxeo-pagination-controls.js` is the tenth and carries **no code change** — see below.

Three test files add 10 tests: 4 on `nuxeo-share-button` `_copyLink`, 4 on `nuxeo-document-acl-table` `_updateAces`, and 2 on the pagination guard.

## Notes

**`_computeLimitForOptions` was deliberately left alone.** Its `S1126` finding is accepted rather than fixed, because every mechanical rewrite diverges on the values the computed binding actually sees. `numberOfPages` is `undefined` before the page provider first binds and can be `NaN`; `NaN > max` is false so the select stays enabled, but `NaN <= max` is false too and would disable it. `!(numberOfPages > max)` preserves behaviour but then reports `S1940`, whose only suggested fix is the broken `<=`. The only edit to that file is a comment recording this, plus two unit tests pinning the `undefined`/`null` and `NaN` cases.

**The `Math.min`/`Math.max` swap cannot diverge.** `Math.min(NaN, n)` is `NaN` where the ternary would have returned the other operand, but both operands are guaranteed numbers by the two `typeof … === 'number'` guards immediately above the changed lines.

**Both `S4043` copies are defensive only.** `labels` and `aces` are both built locally in the same function, so nothing outside could observe the in-place mutation. Ordering of the result is unchanged.

**`S7741` is exact here.** Both sites test a declared binding — a parameter and a property — so `typeof`'s tolerance of undeclared identifiers was never load-bearing, and `null` still takes the same branch as before.

**`indexOf` matches `findIndex` with a `===` predicate**, including returning `-1` when absent. The values are DOM elements, so the `NaN` case where the two differ cannot arise.

**`TypeError` is what `S7786` asks for** on a type-validation failure, and it remains an `instanceof Error`. Nothing in the codebase inspects this error's exact constructor or `name`. The neighbouring `throw` for a missing `entity-type` is left as `Error`, since that is a missing-property failure rather than a type failure.

## Test plan

- CI green — `lint`, `test`, `storybook`, `Build and analyze`, CodeQL and the downstream `web-ui` integration job all pass.
- SonarCloud quality gate passed, 0 new issues, 96.2% coverage on new code. All 11 findings go to 0.
- `npm test` — 4086 passed / 2 skipped on the `maintenance-3.1.x` baseline, plus the 10 new cases.

**One changed line is not covered:** `nuxeo-actions-menu.js:180`. Its enclosing `iron-overlay-closed` handler has no test at all, so this is a pre-existing gap rather than something this PR introduced. Every other changed line is executed, and `nuxeo-user-tag.js` reads 100% line and branch coverage.

The added tests assert behaviour rather than just executing lines — ACE ordering including open-ended entries first, the filtering of denied and archived entries, the share-button copy including its 2-second revert and its failure path, and the two pagination boundary values.

QA is scoped to the one uncovered line and tracked in [ELEMENTS-2094](https://hyland.atlassian.net/browse/ELEMENTS-2094).

### CODE REVIEW

**Comment by naveen-konda** [comment]
## The unit test suite was hanging on this PR

The `test` check on this PR never completed — it ran past the 900s `testsFinishTimeout` reporting `0 passed, 0 failed`. Reproduced locally: `npm run test:ui` sat for 7m40s without emitting a single test result before I killed it.

### Root cause

`ui/nuxeo-slots.js`, `_effectiveParent()`. The refactor changed the returned `parentSlot` from the terminal native `<slot>` to `this` (the `<nuxeo-slot>` element itself):

```js
// before — parentSlot is the terminal native <slot>
parentSlot = this;
while (parentSlot.assignedSlot) {
  parentSlot = parentSlot.assignedSlot;
}
parent = parentSlot.getRootNode().host;
return { parent, parentSlot };

// after — parent is still correct, parentSlot is not
let slot = this.assignedSlot;
while (slot.assignedSlot) {
  slot = slot.assignedSlot;
}
return { parent: slot.getRootNode().host, parentSlot: this };
```

The method's own doc comment states that `parentSlot` is "the native HTML slot in the parent where this element will be displayed", and the consumer in `_render()` relies on exactly that — it reads `parentSlot.getAttribute('name')` and stamps it onto each contributed child as its `slot` attribute.

Returning `this` makes it stamp the **nuxeo-slot registry key** (e.g. `DOCUMENT_ACTIONS`) instead of the **native slot name** (e.g. `header`). Slot-contributed content is therefore never distributed to any native slot and silently disappears; because the registry key gets assigned as a slot name, the re-registration/re-render path no longer terminates and the JS thread blocks before Mocha can even report a test start — hence zero tests reported rather than a normal failure.

Reverting `_effectiveParent()` to the original implementation fixes it. `ui/test/nuxeo-slots.test.js` → "compatibility with native HTML slots" is the suite that pins this behaviour (it asserts `content.assignedSlot` equals `nxSlot.assignedSlot`).

### Also reverted in this push — out of scope for a mechanical maintainability ticket

**1. `_copyLink` in `ui/actions/nuxeo-share-button.js`** — restored `document.execCommand('copy')`.

`navigator.clipboard` is `[SecureContext]`, so on a plain-HTTP deployment the `if (navigator.clipboard && navigator.clipboard.writeText)` guard is false and the fallback path runs `input.select()` — which only highlights the text, copying nothing — and then calls `onCopied()` regardless. The user gets a "Copied" toast and a check icon while the clipboard is untouched. The promise also had no `.catch()`, so a `NotAllowedError` (no document focus / no `clipboard-write` permission) would surface as an unhandled rejection.

**S1874 (`document.execCommand` deprecation) therefore remains OPEN** and needs its own ticket. It is not a mechanical swap: migrating to the async Clipboard API requires a product decision on whether plain-HTTP deployments are still supported, plus a real fallback and error path for the non-secure-context and permission-denied cases.

**2. Restored the JSDoc blocks in `nuxeo-edit-user.html` and `nuxeo-view-user.html`** — the PR deleted the element description, usage example and `@group`/`@element` tags that feed the component docs and Storybook. No Sonar rule asks for documentation removal.

**3. `ui/nuxeo-pagination-controls.js` `_computeLimitForOptions`** — guarded the null case. `numberOfPages > max ? false : true` and `numberOfPages <= max` diverge when `numberOfPages` is `undefined` (before the page provider first binds): the old form returned `true`, the new one returns `false`, which flips the `dom-if` around the page picker. Now `numberOfPages == null || numberOfPages <= max`.

### Kept, but worth a smoke test

`ui/search/nuxeo-search-results-layout.js:153` — replacing `[].concat.apply(children, children.concat(...))` with `children.concat(Array.from(...))` drops a duplicated accumulator and is arguably a genuine bug fix, but it does change the resulting array. Worth exercising the search results layout once before merge.

### Validation

| | maintenance-3.1.x | lts-2025 |
|---|---|---|
| `npm run lint` | exit 0 | exit 0 |
| `npm test` | **terminates**, 4076 passed / 2 skipped | **terminates**, 4075 passed / 2 skipped |

(164 core + 3883 ui + 29 dataviz on `maintenance-3.1.x`; ui is 3882 on `lts-2025` because `nuxeo-uploader-behavior.test.js` has one test fewer there — matching that branch's own green baseline.)

**Comment by naveen-konda** [comment]
### CI follow-up after the fix

**The hang is confirmed gone.** The `test` check is now green on both PRs, and the `Lint` check passes.

Two notes on the remaining red `Build and analyze`:

**1. The first attempt hit an unrelated, non-reproducible stall.** On attempt 1 the `SonarQube` workflow's `npm test` step sat for the full 900s on the ui package with `0 passed`. That was *not* code — the `Test` workflow passed on the exact same commit (`3883 passed, 0 failed, 2 skipped`), and re-running the same job on the same SHA completed the test step normally. Worth knowing that this job can stall spuriously; re-running clears it.

**2. The only genuine remaining failure is the new-code coverage gate**, not a code-quality issue:

| condition | threshold | actual |
|---|---|---|
| `new_coverage` | ≥ 90 % | **89.3 %** ❌ |
| `new_reliability_rating` | A | A ✅ |
| `new_security_rating` | A | A ✅ |
| `new_maintainability_rating` | A | A ✅ |
| `new_duplicated_lines_density` | ≤ 3 % | 0.0 % ✅ |
| `new_security_hotspots_reviewed` | 100 % | 100 % ✅ |

The 0.7 pp shortfall comes from restoring the `_copyLink` body in `nuxeo-share-button.js`: those lines now count as changed code, and `ui/test/nuxeo-share-button.test.js` only exercises `_isAvailable` and `_buildPermalink` — `_copyLink` has never had a test. The `_computeLimitForOptions` null guard is likewise uncovered.

I deliberately did **not** add tests for this in the same pass, to keep the change set limited to the regression fix. Closing the gate is a small, self-contained follow-up: a unit test that stubs `document.execCommand` and asserts `_copyLink` sets the `check` icon / `selected` class and notifies, plus two cases for `_computeLimitForOptions` (`undefined` → `true`, and the `> max` boundary). Happy to add those here if preferred.

**Comment by naveen-konda** [comment]
### Quality gate: new-code coverage

The only failing quality-gate condition was **coverage on new code (89.3% vs a 90% threshold)**. It is now **96.3%**, so the gate passes.

Two things got it there.

**1. Reverted the `_grabResults` change in `ui/search/nuxeo-search-results-layout.js`** (see the deferred-bug note below). The file is now byte-identical to the PR base, so it drops out of the new-code set entirely, taking its one uncovered line with it.

**2. Added 9 unit tests**, which cover the last reachable uncovered line of new code:

| Test file | Tests | Covers |
|---|---|---|
| `ui/test/nuxeo-document-acl-table.test.js` | 4 | `_updateAces` — the `[...aces].sort(...)` line changed by this PR (**this was the remaining uncovered new line**) |
| `ui/test/nuxeo-share-button.test.js` | 4 | `_copyLink` — previously untested end to end |
| `ui/test/nuxeo-pagination-controls.test.js` | 1 | the `numberOfPages == null` guard before the page provider first binds |

`_copyLink` had never had a test: `nuxeo-share-button.js` goes from 8/22 lines and 2/4 branches to **21/22 lines and 4/4 branches**. The new tests stub `document.execCommand` with sinon and assert on observable state — the icon's `icon` property and `selected` class, the native input's selection range, and the `notify` event detail — covering the success path, the 2s debouncer revert (via sinon fake timers) and the failed-copy early return.

Full suite: **4085 passed / 2 skipped** (was 4076 / 2). `npm run lint` exits 0.

---

### Deferred: latent bug in `_grabResults` — needs its own ticket

This PR briefly replaced the descendant-collection expression in `_grabResults` and that has been **reverted** to keep this mechanical maintainability pass behaviour-neutral. The original expression is restored verbatim:

```js
children = [].concat.apply(children, children.concat(dom(el.root).querySelectorAll('*')));
```

**The latent bug:** `Array.prototype.concat` does not spread a `NodeList` — it is not a real array and has no `Symbol.isConcatSpreadable` — so the whole `NodeList` is appended as a *single element* rather than its members being added individually. As a result `_grabResults` never actually collects the queried descendants, and its recursive path is effectively dead code: the recursion re-enters with a `children` array whose entries are `NodeList` objects, not elements, so none of them have a `.root` or match `nuxeo-results`.

**Why it was not fixed here:** the corrected form changes what the layout finds and therefore what it renders, which is out of scope for a SonarCloud clean-up ticket.

**For whoever picks this up:**
- Fixing it needs a smoke test across all search-results layouts — list, grid and table — because it changes which `nuxeo-results` element the layout resolves to.
- The recursion is unbounded in breadth: each level does `querySelectorAll('*')` on every shadow root found, so once the descendants genuinely enter the array the traversal will walk the entire subtree. Depth/size guards are worth considering as part of the fix.

---

### Rule-mislabel audit (S4043, S7766) — no changes needed

The ticket description mislabelled several rules, so the two riskiest sites were re-checked against what the rules actually say. **Both implementations are correct; nothing had to be reverted.**

**S4043** — the ticket said *"add compare function to `Array#sort()`"*, but the rule is *"array-mutating methods should not be used misleadingly"* (do not consume the return value of `sort()` / `reverse()`; copy first). Those are opposite instructions, and following the ticket text would have silently changed sort order. Both sites implement the **real** rule:

- `ui/nuxeo-aggregation/nuxeo-aggregation-behavior.js` (`labelForDirectoryEntry`) — `labels.reverse()` → `labels.slice().reverse()`. Copy-before-mutate, no comparator involved.
- `ui/nuxeo-document-permissions/nuxeo-document-acl-table.js` (`_updateAces`) — `aces.sort(this._sortAces)` → `[...aces].sort(this._sortAces)`. The `_sortAces` comparator was **already there before this PR**; it was not added or modified, so ordering is untouched.

In both cases the array is function-local (`const labels = []` / `const aces = []`) and not referenced after the return, so the copy is a defensive no-op and the resulting order is byte-for-byte what it was before.

**S7766** — the ticket said *"use `.find()` instead of `.filter()[0]`"*, but the rule is *"ternary expressions should be replaced with `Math.min()` / `Math.max()`"*. The diff contains only the Math change, in `_selected` in `ui/nuxeo-page-provider-display-behavior.js`:

```js
const start = Math.min(index, last);   // was: index > last ? last : index
const end   = Math.max(index, last);   // was: index > last ? index : last
```

Equivalent as written: the enclosing block already guards `typeof index === 'number'` and `typeof this._lastSelectedIndex === 'number'`. (The ternary and `Math.min`/`Math.max` only diverge on `NaN` and `-0`, neither of which is reachable for a row index.)

No stray `.find()` conversion was made anywhere in this PR — the diff adds zero `.find(` calls, so nothing was changed on the strength of the wrong description.

---

### Still uncovered on new code (accepted)

`ui/widgets/nuxeo-actions-menu.js:180` — `path = path.slice(0, path.indexOf(action))`. It sits inside the `_actionsMenuReparent` closure that is created inside `idlePeriod.run()` during dialog reparenting, and only runs on an `iron-overlay-closed` event from a reparented action. Reaching it needs a full dialog reparent-and-close cycle; the cost and flakiness of that setup were not worth it with the gate already at 96.3%.

**Comment by naveen-konda** [comment]
### Follow-up: adversarial audit items

Four items were raised. **Three were valid and are fixed; one is a false positive and the change is kept.**

#### 1. `Error` → `TypeError` in `nuxeo-routing-behavior.js` — KEPT, it is in scope

The audit called this discretionary scope creep on the basis that no Sonar rule requires it. That is not correct. **`javascript:S7786` — "Generic `Error` should be `TypeError` when thrown after type checking" — is in this ticket's rule scope**, and the call site is exactly the pattern the rule targets: a `typeof` check followed immediately by a throw.

```js
if (typeof obj !== 'object') {
  throw new TypeError(`cannot resolve route: "${obj}" is not a valid entity object`);
}
```

The rule explicitly lists `typeof` among the type-checking patterns it recognises, and its own guidance is "replace `Error` with `TypeError` when throwing after type checking conditions; the error message can remain the same" — which is precisely what was done. No revert.

#### 2. Pagination guard is now exactly identity — fixed

The audit was right: `numberOfPages == null || numberOfPages <= max` diverged from the original on `NaN`, returning `false` where the original returned `true`. Now:

```js
return !(numberOfPages > maxItemsForNuxeoSelectPagination);
```

This is the direct negation of the original `if (numberOfPages > max) { return false; } return true;`, so it is identical for every input including `undefined`, `null` and `NaN`, and it still satisfies S1126. A test was added pinning the `NaN` case; the existing `undefined` / `null` test remains correct.

#### 3. Nested ternary in `_generateUrl` — fixed

The rewrite had produced a directly nested ternary (a new S3358 finding on a changed line). Restructured to an `if` with an intermediate `const`, which also keeps the original S4624 nested-template-literal fix:

```js
_generateUrl(baseUrl, path) {
  let base = baseUrl;
  if (this.router.useHashbang) {
    const separator = baseUrl.endsWith('/') ? '' : '/';
    base = `${baseUrl}${separator}#!`;
  }
  return `${base}${base.endsWith('/') || path.startsWith('/') ? '' : '/'}${path}`;
}
```

Output verified identical to the **pre-PR original** across all four headline combinations (and 24 combinations in total, including empty base and path):

| `useHashbang` | `baseUrl` | result |
|---|---|---|
| `true` | `/nuxeo/ui` | `/nuxeo/ui/#!/browse` |
| `true` | `/nuxeo/ui/` | `/nuxeo/ui/#!/browse` |
| `false` | `/nuxeo/ui` | `/nuxeo/ui/browse` |
| `false` | `/nuxeo/ui/` | `/nuxeo/ui//browse` |

All four are already pinned by the existing tests in `ui/test/nuxeo-routing-behavior.test.js`, which stay green.

#### 4. `Object.assign` in the new test helper — fixed

The `ace()` fixture in `ui/test/nuxeo-document-acl-table.test.js` now uses object spread. Note it needs an explicit `return` block rather than the concise `=> ({ ... })` form, because this repo configures `arrow-body-style: ['error', 'as-needed', { requireReturnForObjectLiteral: true }]`:

```js
const ace = (id, begin, extra) => {
  return { id, begin, granted: true, status: 'effective', ...extra };
};
```

The four `_updateAces` tests still pass. The six pre-existing test files using `Object.assign` were left untouched as instructed.

---

**Validation:** `npm run lint` exits 0 on both branches. Full suite is **4086 passed / 2 skipped** on maintenance-3.1.x and **4085 / 2** on lts-2025 (one more than before on each, from the new `NaN` case). Local new-code coverage estimate is unchanged at **96.3%**; CI and SonarCloud were not waited on for this round.

Both branches remain byte-identical — all 13 touched files have matching blob hashes across the pair.

File: ui/actions/nuxeo-share-button.js [hunk 1]
```diff
@@ -166,24 +166,27 @@ import '../nuxeo-button-styles.js';
...
+        this.notify({ message: this.i18n('shareButton.operation.copied'), duration: 2000 });
+      };
 
-      // Select Link
-      link.$.paperInput.$.nativeInput.select();
-      if (!window.document.execCommand('copy')) {
+      if (navigator.clipboard && navigator.clipboard.writeText) {
+        navigator.clipboard.writeText(input.value).then(onCopied);
         return;
       }
```
[the remark this answers is not in the document]
Comment (naveen-konda) [reply ui/actions/nuxeo-share-button.js]: Correct, and the change this was written against is gone — the `navigator.clipboard` rewrite was reverted in full, for essentially the reason you gave plus one more.

At the current head (`de8c7e5`), `ui/actions/nuxeo-share-button.js` is byte-identical to `maintenance-3.1.x` — same blob on both sides:

```
$ git rev-parse origin/maintenance-3.1.x:ui/actions/nuxeo-share-button.js
cf8b5a7a21f82be1082e81165d2fe964cb201d50
$ git rev-parse origin/ELEMENTS-2091-maintainability-singleton:ui/actions/nuxeo-share-button.js
cf8b5a7a21f82be1082e81165d2fe964cb201d50
```

It no longer appears in the PR's changed-file list at all; only `ui/test/nuxeo-share-button.test.js` does. `_copyLink()` uses `window.document.execCommand('copy')` again.

The reasoning, for the record: `navigator.clipboard` is `[SecureContext]` in the Clipboard API spec, so the whole object is `undefined` over plain HTTP (anything other than `localhost`) — `document.execCommand('copy')` has no such restriction. On an HTTP deployment the behaviour therefore *inverted*: copy stopped working where it previously worked. And as you spotted, the insecure-context path was worse than a silent failure — it called `input.select()`, which only highlights text, and then ran the success path anyway, showing the check icon and the "Copied" notification with the clipboard untouched. No `.catch()` either, so a `NotAllowedError` (unfocused document, `clipboard-write` denied) gave no feedback.

The S1874 deprecation itself is not dropped: it is recorded in **ELEMENTS-2093 §4** as accept-later, pending a product decision on whether non-secure-context (plain HTTP) deployments are still supported. If they are, `execCommand` has to stay as a real fallback rather than a decorative one, and doing the clipboard path properly needs a failure notification plus tests stubbing both branches — a small feature, not a Sonar sweep. `_copyLink` now has 4 unit tests covering the `execCommand` path (added in this PR), so the clipboard path is the remaining gap.

Resolving — no code change in scope for this PR.

File: ui/nuxeo-slots.js [hunk 2]
```diff
@@ -215,18 +215,14 @@ window.nuxeo.slots.setSharedModel = (model) => {
...
-        }
-        parent = parentSlot.getRootNode().host;
+        return { parent: this.parentNode, parentSlot: null };
+      }
+      let slot = this.assignedSlot;
+      while (slot.assignedSlot) {
+        slot = slot.assignedSlot;
       }
-      return { parent, parentSlot };
+      return { parent: slot.getRootNode().host, parentSlot: this };
```
[the remark this answers is not in the document]
Comment (naveen-konda) [reply ui/nuxeo-slots.js]: Good catch — you identified exactly the right defect, and the consequence was even worse than "may stamp content into the wrong slot". The change has been reverted.

`ui/nuxeo-slots.js` no longer appears in this PR at all; at the current head (`de8c7e5`) it is the same blob as base:

```
$ git rev-parse origin/maintenance-3.1.x:ui/nuxeo-slots.js
2a20c55f2b786e4b7187e4a9b9edbdaaad5d241e
$ git rev-parse origin/ELEMENTS-2091-maintainability-singleton:ui/nuxeo-slots.js
2a20c55f2b786e4b7187e4a9b9edbdaaad5d241e
```

`_effectiveParent()` again walks the `assignedSlot` chain and returns the **terminal native `<slot>`**:

```js
if (!this.assignedSlot) {
  parent = this.parentNode;
} else {
  parentSlot = this;
  while (parentSlot.assignedSlot) {
    parentSlot = parentSlot.assignedSlot;
  }
  parent = parentSlot.getRootNode().host;
}
return { parent, parentSlot };
```

With `parentSlot = this`, `_render()`'s `parentSlot.getAttribute('name')` read the `<nuxeo-slot>` registry key (e.g. `DOCUMENT_ACTIONS`) instead of the native slot name (e.g. `header`), so contributed content was distributed to no slot and silently vanished — no error, no warning. Beyond that, the re-registration/re-render path stopped terminating: the `ui` suite ran past the 900 s `testsFinishTimeout` reporting **0 passed, 0 failed**, blocking the JS thread before Mocha could report a single test start. It presented as a hung CI job rather than a test failure, which is how it was found.

Since `nuxeo-slots` is the Web UI's core extension mechanism, S7740 is recorded in **ELEMENTS-2093 §2** as accept-later with the constraint that any future attempt must keep `parentSlot` as the terminal native slot; `ui/test/nuxeo-slots.test.js` ("compatibility with native HTML slots") is the suite that pins it.

Resolving — the flagged code is no longer in the PR.

**Comment by naveen-konda** [comment]
### `javascript:S1940` on `_computeLimitForOptions()` — accepted as a false positive

The SonarCloud analysis of this PR raises `javascript:S1940` at `ui/nuxeo-pagination-controls.js:243`, *"Use the opposite operator (`<=`) instead."* It is a consequence of the S1126 fix in this PR, which collapsed the original `if`/`return` flow into a single negated expression.

**No code change has been made — the finding is marked Accepted in SonarCloud.**

The suggested rewrite would be a regression. `>` and `<=` are not each other's negation once an operand is not a number, because every relational comparison against `NaN` evaluates to `false`:

| `numberOfPages` | `!(n > max)` — as delivered | `n <= max` — Sonar's suggestion |
| --- | --- | --- |
| `NaN` | `true` | `false` |
| `undefined` (coerces to `NaN`) | `true` | `false` |
| number ≤ max | `true` | `true` |
| number > max | `false` | `false` |

`_computeLimitForOptions()` is a Polymer computed binding, so it runs before the page provider first binds while `numberOfPages` is still `undefined`. The original code — `if (numberOfPages > max) { return false; } return true;` — returned `true` in that state. The negated form preserves that exactly; `<=` would flip it to `false` and change which pagination control renders on first paint.

This divergence was found by audit while preparing this PR and fixed deliberately; a code comment at the site records the reason. Reverting to `<=` would reintroduce it.

Recorded in ELEMENTS-2093 §5 alongside the other accepted findings, with the SonarCloud issue keys and the justification wording.

**Review by madhurkulshrestha-hyland — APPROVED** [review]

**Review by vivekchaturved013 — APPROVED** [review]
